man/eunice.1
.TH MITCH 1 "6 Jan 2010" "MIT Athena"
\" RCSID: $HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/man/mitch.1 $ $Id: mitch.1 3956 2010-01-05 20:56:56Z zacheiss $
.SH NAME
eunice \- examine and modify informtion about printers in Moira
.SH SYNOPSIS
.B eunice printername [options]
.SH DESCRIPTION
.I Eunice
is a tool for updating information about printers in Moira.  It
supports all printer modification commands.

If no options are specified, it will assume \fI-info\fR, which will
result in all information about the specified printer being displayed.

.SH OPTIONS

.IP \fB-info\ \fRor\ \fB-i\fR
Display information about a printer. This is the default mode if no
other options are given. The output is similar to that of moira.

.IP \fB-location\ \fIlocation\ \fRor\ \fB-l\ \fIlocation\fR
Update the location field for the specified printer to \fIlocation\fR.
.IP \fB-contact\ \fIcontact\ \fRor\ \fB-c\ \fIcontact\fR
Update the contact field for the specified printer to \fIcontact\fR.

.IP \fB-hostname\ \fIhostname\ \fRor\ \fB-H\ \fIhostname\fR
Update the hostname for the specified printer to \fIhostname\fR.
.IP \fB-mac\ \fIhwaddr\ \fRor\ \fB-m\ \fIhwaddr\fR
Update the Ethernet hardware address for the printer's hostname record
to \fIhwaddr\fR.  Note that this information is associated with the host
record, not the print queue record.
.IP \fB-remotename\ \fIname\ \fRor\ \fB-r\ \fIname\fR
Update the remote queue name for the specified printer to \fIname\fR.
.IP \fB-duplex\ \fIname\ \fRor\ \fB-d\ \fIname\fR
Update the duplex queue name associated with the specific printer to \fIname\fR.
.IP \fB-printserver\ \fIhostname\ \fRor\ \fB-s\ \fIhostname\fR
Update the print server for the specific printer to \fIhostname\fR.  The
print server may first need to be created via the "Print Servers" menu
in the moira(1) client before queues may be assigned to it.

.IP \fB-type\ \fItype\ \fRor\ \fB-T\ \fItype\fR
Update the type for the specified printer to \fItype\fR.  Common types
include CLUSTER, DORM, PRIVATE, and SAP.
.IP \fB-model\ \fImodel\ \fRor\ \fB-M\ \fImodel\fR
Update the model field for the specified printer to \fImodel\fR.  The
model may first need to be created in the Moira aliases table before it
can be used.

.IP \fB-kerbauth\ \fRor\ \fB-K\fR
Require Kerberos authentication of print requests.
.IP \fB-nokerbauth\ \fRor\ \fB-NK\fR
Do not require Kerberos authentication of print requests.
.IP \fB-banner\ \fRor\ \fB-b\fR
Print header pages by default.  They may still be suppressed by
client configuration.
.IP \fB-nobanner\ \fRor\ \fB-nb\fR
Do not print header pages by default.  They may still be enabled by
client configuration.

.IP \fB-lpcacl\ \fIlist\ \fRor\ \fB-l\ \fIlist\fR
Set the LPC acl for the specified printer to \fIlist\fR.

.IP \fB-lpracl\ \fIlist\ \fRor\ \fB-ac\ \fIlist\fR
Set the restrict list for the specified printer to \fIlist\fR.

.IP \fB-verbose\ \fRor\ \fB-v\fR
Give more information.  With this flag, mitch will display the
queries it is making to the moira server.
.IP \fB-noauth\ \fRor\ \fB-n\fR
Do not attempt to perform Kerberos authentication with the Moira server.
Most lookup operations should be still possible without tickets.
.IP \fB-database\ \fIhost:port\ \fRor\ \fB-db\ \fIhost:port\fR
Use the specified host and port to contact the Moira database instead of
the default server.  Both may be symbolic names or numbers.  If the
port is left off, the default Moira server port will be assumed.  The
database chosen will be the one specified on the command line, specified
in the MOIRASERVER environment variable, the hesiod "moira" sloc entry,
or the compiled in default, in that order or preference.

.IP \fB-create\ \fRor\ \fB-C\fR
This will create the specified printer, with information provided by
other options.
.IP \fB-delete\ \fRor\ \fB-D\fR
This will delete the specified printer.
.IP \fB-rename\ \fInewname\ \fRor\ \fB-R\ \fInewname\fR
This will rename the specified printer to \fInewname\fR.

.SH AUTHORS
Mark Manley, MIT Information Services & Technology.
.br
Garry Zacheiss, MIT Information Services & Technology.
.SH SEE ALSO
moira(1)

.SH DIAGNOSTICS
An exit status of 2 indicates a problem contacting the server. An exit
status of 1 indicates that the arguments could not be parsed or some
query to the moira server returned an error. An exit status of 0
indicates that all operations requested completed successfully.

.SH NOTES
The printer name doesn't actually have to be the first argument, but
if you put it anywhere else, it's easy to get the other arguments in the
wrong order and do something other than what you intended.



debian/changelog
debathena-moira (4.0.0+svn20100225-0debathena2) UNRELEASED; urgency=low

  * There is no "lib" section - it's "libs".

 -- Evan Broder <broder@mit.edu>  Fri, 26 Feb 2010 12:50:19 -0500

debathena-moira (4.0.0+svn20100225-0debathena1) unstable; urgency=low

  * New SVN snapshot.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 15:54:26 -0500

debathena-moira (4.0.0+svn20100223-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Drop patch install-headers, which was incorporated upstream.

 -- Evan Broder <broder@mit.edu>  Thu, 25 Feb 2010 02:13:44 -0500

debathena-moira (4.0.0+svn20100104-0debathena1) unstable; urgency=low

  * New SVN snapshot.
  * Build and install libmrclient as a shared library.
  * Deal with rename of /etc{/athena => }/moira.conf
  * Patch manpages to deal with our installation paths.

 -- Evan Broder <broder@mit.edu>  Mon, 04 Jan 2010 22:05:10 -0500

debathena-moira (4.0.0+cvs20091116-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop patch fix-update-ticket, which was incorporated upstream.
  * Add a missing definition of ${prefix} in include/Makefile.in so that
    the moira.h header actually gets installed in
    /usr/include/moira. (Trac: #445)

 -- Evan Broder <broder@mit.edu>  Thu, 19 Nov 2009 10:53:40 -0500

debathena-moira (4.0.0+cvs20090922-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Add patch to fix builds without krb4.

 -- Evan Broder <broder@mit.edu>  Tue, 22 Sep 2009 00:35:12 -0400

debathena-moira (4.0.0+cvs20090728-0debathena3) unstable; urgency=low

  * Add a missing conflicts to make upgrades run more smoothly.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 18:28:33 -0700

debathena-moira (4.0.0+cvs20090728-0debathena2) unstable; urgency=low

  * Include the headers in debathena-libmoira-dev.
  * Install update_test in moira-clients instead of moira-update-server.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 17:54:34 -0700

debathena-moira (4.0.0+cvs20090728-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Fix zephyr dependencies to actually pull in Kerberized libraries; drop
    build-dep on Kerberized library.

 -- Evan Broder <broder@mit.edu>  Tue, 28 Jul 2009 13:43:28 -0400

debathena-moira (4.0.0+cvs20090621-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Drop krb4less-fix; it was incorporated upstream.
  * Universally stop building Moira with krb4 support.

 -- Evan Broder <broder@mit.edu>  Fri, 26 Jun 2009 00:51:28 -0400

debathena-moira (4.0.0+cvs20090618-0debathena1) unstable; urgency=low

  * New upstream release
  * Build and install libmoira as a shared library. (Trac: #70)
  * Drop most of the krb4 patch - it's been incorporated upstream.
  * Install the Moira development headers by patching the relevant
    Makefiles, instead of in the debian/rules file.

 -- Evan Broder <broder@mit.edu>  Sat, 20 Jun 2009 14:45:32 -0400

debathena-moira (4.0.0+cvs20090409-0debathena1) unstable; urgency=low

  * New CVS snapshot (Trac: #195)
  * Drop patches that have been incorporated upstream.
  * Add patch to allow building on systems without krb4. (Trac: #198)

 -- Evan Broder <broder@mit.edu>  Sun, 12 Apr 2009 14:03:22 -0400

debathena-moira (4.0.0+cvs20080830-0debathena2) unstable; urgency=low

  * Remove DEB_AUTO_UPDATE_DEBIAN_CONTROL.
  * Change "Debian-Athena Project" to "Debathena Project".
  * Remove debian/control from version control.

 -- Tim Abbott <tabbott@mit.edu>  Sat, 24 Jan 2009 15:32:35 -0500

debathena-moira (4.0.0+cvs20080830-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Add a debathena-moira-update-server package that installs
    update_server

 -- Evan Broder <broder@mit.edu>  Fri, 01 Aug 2008 10:54:26 -0400

debathena-moira (4.0.0+cvs20080701-0debathena1) unstable; urgency=low

  * New CVS snapshot
  * Move Moira header files from /usr/share/include to /usr/include/moira

 -- Evan Broder <broder@mit.edu>  Tue, 01 Jul 2008 20:38:13 -0400

debathena-moira (4.0.0+cvs20080304-0debathena1) unstable; urgency=low

  * New CVS snapshot to fix chsh.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 04 Mar 2008 15:50:31 -0500

debathena-moira (4.0.0+cvs20080301-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Remove mrtest segfault patch from last release (fixed in upstream)

 -- Tim Abbott <tabbott@mit.edu>  Sat,  1 Mar 2008 18:51:40 -0500

debathena-moira (4.0.0+cvs20070817-0debathena1) unstable; urgency=low

  * New CVS snapshot.
  * Patch an mrtest segfault with too many command line arguments.

 -- Anders Kaseorg <andersk@mit.edu>  Fri, 24 Aug 2007 19:53:49 -0400

debathena-moira (4.0.0+cvs20070129-0debathena3) unstable; urgency=low

  * Move chsh.debathena to chsh.moira and chfn.debathena to chfn.moira.

 -- Anders Kaseorg <andersk@mit.edu>  Mon, 06 Aug 2007 03:17:10 -0400

debathena-moira (4.0.0+cvs20070129-0debathena2) unstable; urgency=low

  * Add missing build dependencies to libmoira-dev.
  * Section: debathena/net.

 -- Anders Kaseorg <andersk@mit.edu>  Tue, 17 Jul 2007 06:34:07 -0400

debathena-moira (4.0.0+cvs20070129-0debathena1) unstable; urgency=low

  * Initial release.

 -- Anders Kaseorg <andersk@mit.edu>  Wed, 14 Feb 2007 00:53:24 -0500


clients/eunice/eunice.c
/* $Id: eunice.c 3977 2010-02-12 21:12:04Z zacheiss $
 *
 * Command line oriented Moira print queue tool.
 *
 * Code based on the blanche command line tool and the moira curses tool.
 * Copyright (C) 2009 by the Massachusetts Institute of Technology.
 * For copying and distribution information, please see the file
 * <mit-copyright.h>.
 * 
 * Mark Manley, mmanley@MIT.EDU, 12/09
 */

#include <mit-copyright.h>
#include <moira.h>
#include <moira_site.h>
#include <mrclient.h>

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

RCSID("$HeadURL: svn+ssh://svn.mit.edu/moira/trunk/moira/clients/eunice/eunice.c $ $Id: eunice.c 3977 2010-02-12 21:12:04Z zacheiss $");

struct member {
  int type;
  char *name, *tag;
};

const char *deflogserver = "WSLOGGER.MIT.EDU";

/* legacy variables that we need to set regardless */
const char *quotaserver = "[NONE]";
const char *pageprice = "10";

/* argument parsing macro */
#define argis(a, b) (!strcmp(*arg + 1, a) || !strcmp(*arg + 1, b))

/* flags from command line */
int info_flag, verbose, noauth, duplex_default, hold_default;
int create_flag, setmac, delete_flag, rename_flag, ka, banner, update_flag;
char *lpracl, *lpcacl, *report_list;
char *contact, *newname, *printserver, *type, *hwtype, *mac, *hostname, *queuename;
char *duplexname, *logserver, *location, *realname, *pr_status;

char *queuename, *whoami, *testqueue;

void usage(char **argv);
int show_printer_info(char *queuename);
int save_printer_info(int argc, char **argv, void *hint);
int save_hwaddr(int argc, char **argv, void *hint);
void recursive_display_list_members(void);
char *get_username(void);
int wrap_mr_query(char *handle, int argc, char **argv,
		  int (*callback)(int, char **, void *), void *callarg);
void print_query(char *query_name, int argc, char **argv);
int CountArgs(char **info);

static char *states[] = {
  "Reserved (0)",
  "Active (1)",
  "None (2)",
  "Deleted (3)"
};

static char *PrnState(int state)
{
  static char buf[BUFSIZ];
  
  if (state < 0 || state > 3)
    {
      sprintf(buf, "Unknown (%d)", state);
      return buf;
    }
  return states[state];
}

int main(int argc, char **argv)
{
  int status, success;
  char **arg = argv;
  char *uargv[2];
  char *pargv[PRN_END + 1];
  char *membervec[4];
  struct member *memberstruct;
  char *server = NULL, *p;
  int i;

  /* clear all flags & lists */
  i = info_flag = verbose = noauth = 0;
  setmac = create_flag = delete_flag = update_flag = 0;
  ka = duplex_default = hold_default = banner = -1;
  location = lpracl = lpcacl = report_list = pr_status = NULL;
  logserver = duplexname = realname = newname = printserver = type = hwtype = mac = hostname = NULL;


  contact = NULL;
  whoami = argv[0];

  success = 1;

  /* parse args */
  while (++arg - argv < argc)
    {
      if (**arg == '-')
	{
	  if (argis("i", "info"))
	    info_flag++;
	  else if (argis("C", "create"))
	    create_flag++;
	  else if (argis("D", "delete"))
	    delete_flag++;
	  else if (argis("L", "location"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  location = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("r", "remotename"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  realname = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("T", "type"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  type = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("s", "printserver"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  printserver = canonicalize_hostname(strdup(*arg));
		}
	      else
		usage(argv);
	    }
	  else if (argis("M", "model"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  hwtype = *arg;
		}
	      else
		usage(argv);
	    }
          else if (argis("S", "status"))
	    {
	      if (arg - argv < argc - 1) {
		int i;
		int len;
		
		arg++;
		update_flag++;
		pr_status = *arg;
		
		len = strlen(pr_status);
		for(i = 0; i < len; i++) {
		  if(!isdigit(pr_status[i])) {
		    printf("Error: status code %s is not numeric.\n", pr_status);
		    exit(1);
		  }
		}
	      } else
		usage(argv);
	    }
	  else if (argis("K", "kerbauth")) 
 	    {
	      ka = 1;
  	      update_flag++;
 	    }
	  else if (argis("NK", "nokerbauth"))
 	    {
	      ka = 0;
  	      update_flag++;
 	    }
	  else if (argis("DD", "duplexdefault"))
	    {
	      duplex_default = 1;
	      update_flag++;
	    }
	  else if (argis("ND", "noduplexdefault"))
	    {
	      duplex_default = 0;
	      update_flag++;
	    }
	  else if (argis("HD", "holddefault"))
	    {
	      hold_default = 1;
	      update_flag++;
	    }
	  else if (argis("NH", "noholddefault"))
	    {
	      hold_default = 0;
	      update_flag++;
	    }
	  else if (argis("H", "hostname"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
  		  update_flag++;
		  hostname = canonicalize_hostname(strdup(*arg));
		}
	      else
		usage(argv);
	    }
	  else if (argis("n", "noauth"))
	    noauth++;
	  else if (argis("v", "verbose"))
	    verbose++;
	  else if (argis("S", "server") || argis("db", "database"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  server = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("c", "contact"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  contact = *arg;
		  update_flag++;
		}
	      else
		usage(argv);
	    }
	  else if (argis("l", "lpcacl"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  lpcacl = *arg;
		  update_flag++;
		}
	      else
		usage(argv);
	    }
	  else if (argis("ac", "lpracl"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  lpracl = *arg;
		  update_flag++;
		}
	      else
		usage(argv);
	    }
	  else if (argis("rl", "reportlist"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  report_list = *arg;
		  update_flag++;
		}
	      else
		usage(argv);
	    }
	  else if (argis("m", "mac"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
		  setmac++;
		  mac = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("b", "banner"))
	    {
	      update_flag++;
	      banner = 1;
	    }
	  else if (argis("nb", "nobanner"))
	    {
	      update_flag++;
	      banner = 0;
	    }
	  else if (argis("L", "logserver"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
	          update_flag++;
		  logserver = canonicalize_hostname(strdup(*arg));
		}
	      else
		usage(argv);
	    }
	  else if (argis("R", "rename"))
	    {
	      if (arg - argv < argc - 1)
		{
		  rename_flag++;
		  update_flag++;
		  ++arg;
		  newname = *arg;
		}
	      else
		usage(argv);
	    }
	  else if (argis("d", "duplex"))
	    {
	      if (arg - argv < argc - 1)
		{
		  ++arg;
	          update_flag++;
		  duplexname = *arg;
		}
	      else
		usage(argv);
	    }
	  else
	    usage(argv);
	}
      else if (queuename == NULL)
	queuename = *arg;
      else
	usage(argv);
    }
  if (queuename == NULL)
    usage(argv);


  if (!update_flag && !rename_flag && !delete_flag && !create_flag && !setmac)
    info_flag++;

  /* fire up Moira */
  status = mrcl_connect(server, "eunice", 13, !noauth);
  if (status == MRCL_AUTH_ERROR)
    {
      com_err(whoami, 0, "Authentication error while working on queue %s",
	      queuename);
      com_err(whoami, 0, "Try the -noauth flag if you don't "
	      "need authentication.");
    }
  if (status)
    exit(2);

  if (create_flag)
    {
      if (hostname == NULL || type == NULL || hwtype == NULL )
        usage(argv);
    }

  for (i = 0; i < PRN_END + 1; i++)
      pargv[i] = NULL;

  /* check for name conflicts. */
  if (create_flag || rename_flag)
    {

      if (rename_flag)
	testqueue = newname;
      else
	testqueue = queuename;

      status = wrap_mr_query("get_printer", 1, &testqueue, NULL, NULL);
      if (status != MR_NO_MATCH) 
      {
	fprintf(stderr, "ERROR: A queue by that name already exists.\n");
	exit(1);
      }

      status = wrap_mr_query("get_printer_by_duplexname", 1, &testqueue, NULL, NULL);
      if (status != MR_NO_MATCH)
      {
	fprintf(stderr, "ERROR: A duplex queue by that name already exists.\n");
	exit(1);
      }

      if (duplexname) {
        status = wrap_mr_query("get_printer_by_duplexname", 1, &testqueue, NULL, NULL);
        if (status != MR_NO_MATCH)
	{
	  fprintf(stderr, "ERROR: A duplex queue by that name already exists.\n");
	  exit(1);
	}
      }

    }

  /* create if needed */
  if (create_flag)
    {
      if (realname == NULL)
        realname = queuename;	/* Variable for aliases */

      if (logserver == NULL)
        logserver = (char *) canonicalize_hostname(strdup(deflogserver));

      pargv[PRN_NAME] = queuename;
      pargv[PRN_TYPE] = type;
      pargv[PRN_HWTYPE] = hwtype;

      pargv[PRN_DUPLEXNAME] = (duplexname != NULL) ? duplexname : "";

      pargv[PRN_DUPLEXDEFAULT] = (duplex_default == 1) ? "1" : "0";
      pargv[PRN_HOLDDEFAULT] = (hold_default == 1) ? "1" : "0";

      if (pr_status)
	pargv[PRN_STATUS] = pr_status;
      else
	pargv[PRN_STATUS] = "1";

      pargv[PRN_HOSTNAME] = hostname;
      pargv[PRN_LOGHOST] = logserver;

      pargv[PRN_RM] = (printserver != NULL) ? printserver : "[ANY]";

      pargv[PRN_CONTACT] = (contact != NULL) ? contact : "";

      pargv[PRN_LOCATION] = (location != NULL) ? location : "";

      pargv[PRN_RP] = realname;
      pargv[PRN_RQ] = (char *) quotaserver;
      pargv[PRN_KA] = (ka == 1) ? "1" : "0";
      pargv[PRN_PC] = (char *) pageprice;

      if (lpracl != NULL) 
        {
          status = wrap_mr_query("get_list_info", 1, &lpracl,
                        NULL, NULL);
          if (status)
            {
               com_err(whoami, status, "while getting authentication list information");
               exit(1);
            }
          pargv[PRN_AC] = lpracl;
        }
      else
        pargv[PRN_AC] = "[none]";

      if (lpcacl != NULL) 
        {
          status = wrap_mr_query("get_list_info", 1, &lpcacl,
                        NULL, NULL);
          if (status)
            {
               com_err(whoami, status, "while getting authentication list information");
               exit(1);
            }
          pargv[PRN_LPC_ACL] = lpcacl;
        }
      else
        pargv[PRN_LPC_ACL] = "[none]";

      if (report_list != NULL)
        {
          status = wrap_mr_query("get_list_info", 1, &report_list,
				 NULL, NULL);
          if (status)
            {
	      com_err(whoami, status, "while getting authentication list information");
	      exit(1);
            }
          pargv[PRN_REPORT_LIST] = report_list;
        }
      else
        pargv[PRN_REPORT_LIST] = "[none]";

      pargv[PRN_BANNER] = (banner == 1) ? "1" : "0";

      status = wrap_mr_query("add_printer", CountArgs(pargv), pargv, NULL, NULL); 

      if (status)
	{
	  com_err(whoami, status, "while creating print queue.");
	  exit(1);
	}
      else
 	  printf ("%s: queue %s created.\n", whoami, queuename);
    }
  else if (update_flag)
    {

      status = wrap_mr_query("get_printer", 1, &queuename, save_printer_info, pargv);
      if (status)
	{
	  com_err(whoami, status, "while getting printer information");
	  exit(1);
	}

      pargv[0] = queuename;
      if (newname && rename_flag)
	pargv[PRN_NAME + 1] = newname;
      if (type != NULL)
	pargv[PRN_TYPE + 1] = type;
      if (hwtype != NULL)
	pargv[PRN_HWTYPE + 1] = hwtype;

      if (duplexname != NULL)
	pargv[PRN_DUPLEXNAME + 1] = duplexname;

      if (duplex_default != -1)
	pargv[PRN_DUPLEXDEFAULT + 1] = duplex_default ? "1" : "0";

      if (hold_default != -1)
	pargv[PRN_HOLDDEFAULT + 1] = hold_default ? "1" : "0";

      if (pr_status != NULL)
	pargv[PRN_STATUS + 1] = pr_status;

      if (hostname != NULL)
	pargv[PRN_HOSTNAME + 1] = hostname;

      if (logserver != NULL)
	pargv[PRN_LOGHOST + 1] = logserver;

      if (printserver != NULL)
	pargv[PRN_RM + 1] = printserver;

      if (realname != NULL)
	pargv[PRN_RP + 1] = realname;

      if (quotaserver != NULL && strcmp(quotaserver,"[NONE]"))
	pargv[PRN_RQ + 1] = (char *) quotaserver;

      if (ka != -1)
	pargv[PRN_KA + 1] = ka ? "1" : "0";
      if (lpracl != NULL)
	pargv[PRN_AC + 1] = lpracl;
      if (lpcacl != NULL)
	pargv[PRN_LPC_ACL + 1] = lpcacl;
      if (report_list != NULL)
	pargv[PRN_REPORT_LIST + 1] = report_list;
      if (banner != -1)
	pargv[PRN_BANNER + 1] = banner ? "1" : "0";
      if (location != NULL)
	pargv[PRN_LOCATION + 1] = location;
      if (contact != NULL)
	pargv[PRN_CONTACT + 1] = contact;

       pargv[PRN_MODBY + 1] = pargv[PRN_MODTIME + 1] = pargv[PRN_MODWITH + 1] = NULL;

      status = wrap_mr_query("update_printer", CountArgs(pargv), pargv, NULL, NULL);
      if (status)
	{
	  com_err(whoami, status, "while updating print queue.");
	  exit(1);
	}
       else
         {
	   if (rename_flag) 
 	     printf ("%s: queue %s renamed to %s.\n", whoami, queuename, newname);
	   else
 	     printf ("%s: queue %s updated.\n", whoami, queuename);
         }
    }

  /* display printer info if requested to */
  if (info_flag)
    show_printer_info(queuename);

  if (setmac)
    {
      status = wrap_mr_query("get_printer", 1, &queuename, save_printer_info, pargv);
      if (status)
	{
	  com_err(whoami, status, "while getting printer information");
	  exit(1);
	}

      if (hostname == NULL)
	uargv[0] = (char *) strdup (pargv[PRN_HOSTNAME + 1]);
      else
        uargv[0] = (char *) strdup (hostname);
      uargv[1] = (char *) strdup (mac);
      if ((status = wrap_mr_query("update_host_hwaddr", 2, uargv, NULL, NULL)))
        com_err(whoami, status, "updating ethernet address.");
    }


  if (delete_flag)
    {
      status = wrap_mr_query("delete_printer", 1, &queuename,
			NULL, NULL);
      if (status)
	{
	  com_err(whoami, status, "while deleting printer");
	  exit(1);
	}
       else
 	  printf ("%s: queue %s deleted.\n", whoami, queuename);
    }

  /* We're done! */
  mr_disconnect();
  exit(success ? 0 : 1);
}

void usage(char **argv)
{
#define USAGE_OPTIONS_FORMAT "  %-39s%s\n"
  fprintf(stderr, "Usage: %s queue [options]\n", argv[0]);
  fprintf(stderr, "Options are\n");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-v  | -verbose",
	  "-C   | -create");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-D  | -delete",
	  "-R   | -rename newname");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-i  | -info",
	  "-L   | -location location");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-H  | -hostname host",
	  "-c   | -contact contact");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-r  | -remotename name",
	  "-T   | -type printer_type");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-M  | -model model",
	  "-s   | -printserver print_server");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-K  | -kerbauth",
	  "-b   | -banner");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-NK | -nokerbauth",
	  "-nb  | -nobanner");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-DD | -duplexdefault",
	  "-HD | -holddefault");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-ND | -noduplexdefault",
	  "-NH | -noholddefault");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-S | -status status",
	  "-rl | -reportlist list");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-l  | -lpcacl list",
	  "-ac  | -lpracl list");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-d  | -duplex name",
	  "-n   | -noauth");
  fprintf(stderr, USAGE_OPTIONS_FORMAT, "-m  | -mac hwaddr", 
	  "-db  | -database host[:port]");
  exit(1);
}

int show_printer_info(char *queuename)
{
  char hwaddr[20];
  char *pargv[PRN_END + 1];
  int status, banner, i;

  for (i = 0; i < PRN_END + 1; i++)
    pargv[i] = NULL;

  memset (hwaddr,'\0',sizeof(hwaddr));

  status = wrap_mr_query("get_printer", 1, &queuename, save_printer_info, pargv);
  if (status)
    {
      com_err(whoami, status, "while getting printer information");
      exit (1);
    }

  status = wrap_mr_query("get_host_hwaddr", 1, &pargv[PRN_HOSTNAME + 1], 
    save_hwaddr, &hwaddr);

  if (status)
    sprintf (hwaddr,"none");

  banner = atoi(pargv[PRN_BANNER + 1]);

  printf("Printer: %-18s Duplex queue: %-18s\n", pargv[PRN_NAME + 1],
          *pargv[PRN_DUPLEXNAME + 1] ? pargv[PRN_DUPLEXNAME + 1] : "[none]");
  printf("Status: %-10s\n", PrnState(atoi(pargv[PRN_STATUS + 1])));
  printf("Duplex by Default: %-8s Hold by Default: %-18s\n",
	 atoi(pargv[PRN_DUPLEXDEFAULT + 1]) ? "yes" : "no",
	 atoi(pargv[PRN_HOLDDEFAULT + 1]) ? "yes" : "no");
  printf("Type: %-10s Hardware type: %-10s Hardware address: %s\n",
          pargv[PRN_TYPE + 1], pargv[PRN_HWTYPE + 1], hwaddr);
  printf("Printer hostname: %s\n", pargv[PRN_HOSTNAME + 1]);
  printf("Printer log host: %s\n", pargv[PRN_LOGHOST + 1]);
  printf("Spool host: %s\n", pargv[PRN_RM + 1]);
  printf("Remote Printer Name: %-10s Banner page: %s\n", pargv[PRN_RP + 1],
          banner ? ( banner == PRN_BANNER_FIRST ? "Yes" : "Last" ) : "No");
  printf("Authentication: %-3s Price/page: %-3s  Quota Server: %s\n",
          atoi(pargv[PRN_KA + 1]) ? "yes" : "no", pargv[PRN_PC + 1], pargv[PRN_RQ + 1]);
  printf("Restrict list: %-23s  LPC ACL: %-23s\n",
          pargv[PRN_AC + 1], pargv[PRN_LPC_ACL + 1]);
  printf("Report list: %-23s\n", pargv[PRN_REPORT_LIST + 1]);
  printf("Location: %s\n", pargv[PRN_LOCATION + 1]);
  printf("Contact: %s\n", pargv[PRN_CONTACT + 1]);
  printf("\n");
  printf("Last mod by %s at %s with %s.\n", pargv[PRN_MODBY + 1], pargv[PRN_MODTIME + 1], pargv[PRN_MODWITH + 1]);

  return MR_CONT;
}


/* Copy retrieved information about a printer into a new argv */

int save_printer_info(int argc, char **argv, void *hint)
{
  char **nargv = hint;

  for (argc = 0; argc < PRN_END; argc++)
    nargv[argc + 1] = strdup(argv[argc]);
  return MR_CONT;
}

int save_hwaddr(int argc, char **argv, void *hint)
{
  char *p = hint;
  strcpy(p,argv[0]);
  return MR_CONT;
}

int wrap_mr_query(char *handle, int argc, char **argv,
		  int (*callback)(int, char **, void *), void *callarg) 
{
  int status = 0;

  if (verbose)
    print_query(handle, argc, argv);

  status = mr_query(handle, argc, argv, callback, callarg);

  return (status);
}

void print_query(char *query_name, int argc, char **argv)
{ 
  int cnt;

  printf("qy %s", query_name);
  for(cnt=0; cnt<argc; cnt++)
    printf(" <%s>", argv[cnt]);
  printf("\n");
}

char *get_username(void)
{
  char *username;

  username = getenv("USER");
  if (!username)
    {
      username = mrcl_krb_user();
      if (!username)
	{
	  com_err(whoami, 0, "Could not determine username");
	  exit(1);
	}
    }
  return username;
}

/*      Function Name: CountArgs
 *      Description:  Retrieve the number of args in a null terminated
 *                     arglist.
 *      Arguments: info - the argument list.
 *      Returns: number if args in the list.
 */

int CountArgs(char **info)
{
  int number = 0;

  while (*info)
    {
      number++;
      info++;
    }

  return number;
}
